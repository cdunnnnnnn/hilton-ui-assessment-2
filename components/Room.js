import React from 'react'

const Room = ({
  number,
  room: { adults, children, selected },
  handleChange,
  updateRooms,
}) => {
  const disabled = number === 0 ? false : !selected

  return (
    <div
      disabled={disabled}
      className={
        'w-full h-full md:w-1/2 lg:w-1/4 py-4 px-2  mb-6 md:mb-0' +
        (disabled ? ' cursor-not-allowed' : '')
      }
    >
      <div
        className={
          'w-full h-full rounded p-3' +
          (disabled ? ' bg-gray-500' : ' bg-gray-800')
        }
      >
        {number < 1 ? (
          <label className="block uppercase tracking-wide text-gray-500 text-lg font-bold mb-2">
            Room 1
          </label>
        ) : (
          <label
            htmlFor={`selected-${number + 1}`}
            className={
              'block uppercase tracking-wide text-gray-500 text-lg font-bold mb-2' +
              (disabled ? ' text-gray-600' : '')
            }
          >
            <input
              type="checkbox"
              selected
              id={`selected-${number + 1}`}
              className="mr-2 leading-tight"
              checked={selected}
              onChange={event => handleChange(number, event)}
            />
            Room {number + 1}
          </label>
        )}
        <fieldset disabled={disabled} aria-disabled={disabled}>
          <label
            htmlFor={`adults-${number + 1}`}
            className={
              'block uppercase tracking-wide text-xs font-bold mb-2' +
              (disabled
                ? ' cursor-not-allowed text-gray-300'
                : ' text-gray-200')
            }
          >
            Adults (18+)
          </label>
          <div className="relative pb-3">
            <select
              name="adults"
              id={`adults-${number + 1}`}
              className={
                'block appearance-none w-full bg-gray-200 border border-gray-200 py-3 px-4 pr-8 rounded leading-tight focus:outline-none focus:bg-white focus:border-gray-500' +
                (disabled
                  ? ' cursor-not-allowed text-gray-500'
                  : ' text-gray-700')
              }
              value={adults}
              onChange={event => updateRooms(event, 'adults', number)}
            >
              <option value="1">1</option>
              <option value="2">2</option>
            </select>
          </div>
          <label
            htmlFor={`children-${number + 1}`}
            className={
              'block uppercase tracking-wide text-xs font-bold mb-2' +
              (disabled
                ? ' cursor-not-allowed text-gray-300'
                : ' text-gray-200')
            }
          >
            Children (0-17)
          </label>
          <div className="relative pb-3">
            <select
              name="children"
              id={`children-${number + 1}`}
              className={
                'block appearance-none w-full bg-gray-200 border border-gray-200 py-3 px-4 pr-8 rounded leading-tight focus:outline-none focus:bg-white focus:border-gray-500' +
                (disabled
                  ? ' cursor-not-allowed text-gray-500'
                  : ' text-gray-700')
              }
              value={children}
              onChange={event => updateRooms(event, 'children', number)}
            >
              <option value="0">0</option>
              <option value="1">1</option>
              <option value="2">2</option>
            </select>
          </div>
        </fieldset>
      </div>
    </div>
  )
}

export default Room
