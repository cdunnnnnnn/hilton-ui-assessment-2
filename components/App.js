import React, { useEffect, useState } from 'react'

import Room from './Room'

const App = () => {
  const [rooms, setRooms] = useState([
    {
      _id: 1,
      adults: 1,
      children: 0,
      selected: true,
    },
    {
      _id: 2,
      adults: 1,
      children: 0,
      selected: false,
    },
    {
      _id: 3,
      adults: 1,
      children: 0,
      selected: false,
    },
    {
      _id: 4,
      adults: 1,
      children: 0,
      selected: false,
    },
  ])
  const [isLoading, setIsLoading] = useState(true)

  useEffect(() => {
    if (!testStorage()) return false

    if (localStorage['rooms']) {
      const data = JSON.parse(localStorage.getItem('rooms'))
      setRooms(data)
    }
  }, [])

  const testStorage = () => {
    const test = 'test'

    try {
      localStorage.setItem(test, test)
      localStorage.removeItem(test)
      return true
    } catch (error) {
      console.error('localStorage is not available.')
      return false
    }
  }

  const handleChange = (roomNumber, event) => {
    const roomsState = [...rooms]
    const isCheckbox = event.target.type === 'checkbox'
    const checkedRoom = roomsState[roomNumber]
    const boolean = !checkedRoom.selected
    let i = 3

    if (isCheckbox) {
      while (i >= 1) {
        if (boolean) {
          if (i <= roomNumber) {
            roomsState[i].selected = true
          }
        } else {
          if (i >= roomNumber) {
            roomsState[i].selected = false
            roomsState[i].adults = 1
            roomsState[i].children = 0
          }
        }
        i--
      }
      setRooms(roomsState)
      handleSubmit()
    }
  }

  const updateRooms = (event, type, roomNumber) => {
    const roomsState = [...rooms]
    roomsState[roomNumber][type] = event.target.value
    setRooms(roomsState)
  }

  const handleSubmit = () => {
    if (!testStorage()) return false

    localStorage.setItem('rooms', JSON.stringify(rooms))
  }

  return (
    <>
      <form method="post" className="w-full max-w-6xl mx-auto">
        <div className="flex flex-wrap items-center justify-start bg-gray-400 rounded px-2 -mx-3 my-6">
          {rooms.map((room, index) => {
            return (
              <Room
                key={index}
                room={room}
                number={index}
                handleChange={handleChange}
                updateRooms={updateRooms}
              />
            )
          })}
        </div>
        <button
          className="bg-transparent hover:bg-indigo-600 border-4 border-indigo-600 hover:border-indigo-600 text-indigo-600 hover:text-white text-lg font-bold uppercase tracking-wide py-3 px-12 -ml-4 rounded"
          type="submit"
          onClick={() => handleSubmit()}
        >
          Submit
        </button>
      </form>
    </>
  )
}

export default App
