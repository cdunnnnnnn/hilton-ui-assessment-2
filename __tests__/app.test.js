import { mount } from 'enzyme'
import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'

import App from '../components/App'
import Room from '../components/Room'

test('renders the page', () => {
  const div = document.createElement('div')

  ReactDOM.render(<App />, div)
  ReactDOM.unmountComponentAtNode(div)
})

test('matches snapshot of a selectable room', () => {
  const room = {
    adults: 1,
    children: 0,
    selected: false,
  }

  const component = renderer.create(
    <Room
      key={1}
      room={room}
      number={1}
      handleChange={() => {}}
      updateRooms={() => {}}
    />
  )

  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('matches snapshot of room number 1, which is always selected abnd cannot be unselected', () => {
  const room = {
    adults: 1,
    children: 0,
    selected: true,
  }

  const component = renderer.create(
    <Room
      key={0}
      room={room}
      number={0}
      handleChange={() => {}}
      updateRooms={() => {}}
    />
  )

  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('updates a selected room', () => {
  let isSelected = false
  const room = {
    adults: 1,
    children: 0,
    selected: isSelected,
  }

  const component = mount(
    <Room
      key={1}
      room={room}
      number={1}
      handleChange={(roomNumber, event) => {
        isSelected = true
      }}
      updateRooms={() => {}}
    />
  )

  const roomCheckbox = component.find('input[type="checkbox"]')

  roomCheckbox.simulate('change', { target: { checked: true } })
  expect(isSelected).toEqual(true)
})

test('updates an unselected room', () => {
  let isSelected = true
  const room = {
    adults: 1,
    children: 0,
    selected: isSelected,
  }

  const component = mount(
    <Room
      key={1}
      room={room}
      number={1}
      handleChange={(roomNumber, event) => {
        isSelected = false
      }}
      updateRooms={() => {}}
    />
  )

  const roomCheckbox = component.find('input[type="checkbox"]')

  roomCheckbox.simulate('change', { target: { checked: false } })
  expect(isSelected).toEqual(false)
})

test('updates the selected room adults count', () => {
  let adultCount = 1
  const room = {
    adults: adultCount,
    children: 0,
    selected: true,
  }

  const component = mount(
    <Room
      key={1}
      room={room}
      number={1}
      handleChange={() => {}}
      updateRooms={(event, type, roomNumber) => {
        adultCount = event.target.value
      }}
    />
  )

  const adultSelect = component.find('select[name="adults"]')

  adultSelect.simulate('change', { target: { value: 2 } })
  expect(adultCount).toEqual(2)
})

test('updates the selected room children count', () => {
  let childrenCount = 0
  const room = {
    adults: 1,
    children: childrenCount,
    selected: true,
  }

  const component = mount(
    <Room
      key={1}
      room={room}
      number={1}
      handleChange={() => {}}
      updateRooms={(event, type, roomNumber) => {
        childrenCount = event.target.value
      }}
    />
  )

  const childrenSelect = component.find('select[name="children"]')

  childrenSelect.simulate('change', { target: { value: 1 } })
  expect(childrenCount).toEqual(1)
})
