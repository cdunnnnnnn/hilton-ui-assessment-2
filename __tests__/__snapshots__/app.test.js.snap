// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot of a selectable room 1`] = `
<div
  className="w-full h-full md:w-1/2 lg:w-1/4 py-4 px-2  mb-6 md:mb-0 cursor-not-allowed"
  disabled={true}
>
  <div
    className="w-full h-full rounded p-3 bg-gray-500"
  >
    <label
      className="block uppercase tracking-wide text-gray-500 text-lg font-bold mb-2 text-gray-600"
      htmlFor="selected-2"
    >
      <input
        checked={false}
        className="mr-2 leading-tight"
        id="selected-2"
        onChange={[Function]}
        selected={true}
        type="checkbox"
      />
      Room 
      2
    </label>
    <fieldset
      aria-disabled={true}
      disabled={true}
    >
      <label
        className="block uppercase tracking-wide text-xs font-bold mb-2 cursor-not-allowed text-gray-300"
        htmlFor="adults-2"
      >
        Adults (18+)
      </label>
      <div
        className="relative pb-3"
      >
        <select
          className="block appearance-none w-full bg-gray-200 border border-gray-200 py-3 px-4 pr-8 rounded leading-tight focus:outline-none focus:bg-white focus:border-gray-500 cursor-not-allowed text-gray-500"
          id="adults-2"
          name="adults"
          onChange={[Function]}
          value={1}
        >
          <option
            value="1"
          >
            1
          </option>
          <option
            value="2"
          >
            2
          </option>
        </select>
      </div>
      <label
        className="block uppercase tracking-wide text-xs font-bold mb-2 cursor-not-allowed text-gray-300"
        htmlFor="children-2"
      >
        Children (0-17)
      </label>
      <div
        className="relative pb-3"
      >
        <select
          className="block appearance-none w-full bg-gray-200 border border-gray-200 py-3 px-4 pr-8 rounded leading-tight focus:outline-none focus:bg-white focus:border-gray-500 cursor-not-allowed text-gray-500"
          id="children-2"
          name="children"
          onChange={[Function]}
          value={0}
        >
          <option
            value="0"
          >
            0
          </option>
          <option
            value="1"
          >
            1
          </option>
          <option
            value="2"
          >
            2
          </option>
        </select>
      </div>
    </fieldset>
  </div>
</div>
`;

exports[`matches snapshot of room number 1, which is always selected abnd cannot be unselected 1`] = `
<div
  className="w-full h-full md:w-1/2 lg:w-1/4 py-4 px-2  mb-6 md:mb-0"
  disabled={false}
>
  <div
    className="w-full h-full rounded p-3 bg-gray-800"
  >
    <label
      className="block uppercase tracking-wide text-gray-500 text-lg font-bold mb-2"
    >
      Room 1
    </label>
    <fieldset
      aria-disabled={false}
      disabled={false}
    >
      <label
        className="block uppercase tracking-wide text-xs font-bold mb-2 text-gray-200"
        htmlFor="adults-1"
      >
        Adults (18+)
      </label>
      <div
        className="relative pb-3"
      >
        <select
          className="block appearance-none w-full bg-gray-200 border border-gray-200 py-3 px-4 pr-8 rounded leading-tight focus:outline-none focus:bg-white focus:border-gray-500 text-gray-700"
          id="adults-1"
          name="adults"
          onChange={[Function]}
          value={1}
        >
          <option
            value="1"
          >
            1
          </option>
          <option
            value="2"
          >
            2
          </option>
        </select>
      </div>
      <label
        className="block uppercase tracking-wide text-xs font-bold mb-2 text-gray-200"
        htmlFor="children-1"
      >
        Children (0-17)
      </label>
      <div
        className="relative pb-3"
      >
        <select
          className="block appearance-none w-full bg-gray-200 border border-gray-200 py-3 px-4 pr-8 rounded leading-tight focus:outline-none focus:bg-white focus:border-gray-500 text-gray-700"
          id="children-1"
          name="children"
          onChange={[Function]}
          value={0}
        >
          <option
            value="0"
          >
            0
          </option>
          <option
            value="1"
          >
            1
          </option>
          <option
            value="2"
          >
            2
          </option>
        </select>
      </div>
    </fieldset>
  </div>
</div>
`;
